overlord/devicestate/handlers_serial.go

```diff
@@ -335,19 +335,20 @@ func prepareSerialRequest(t *state.Task, regCtx registrationContext, privKey ass
 
 	resp, err := client.Do(req)
 	if err != nil {
-		if !httputil.ShouldRetryError(err) {
-			// a non temporary net error fully errors out and triggers a retry
-			// retries
-			return "", fmt.Errorf("cannot retrieve request-id for making a request for a serial: %v", err)
-		}
-		if httputil.NoNetwork(err) {
-			// If there is no network there is no need to count
-			// this as a tentatives attempt. If we do it this
-			// way the risk is that we tried a bunch of times
-			// with no network and if we hit the server for real
-			// and it replies with something we need to retry
-			// we will not because nTentatives is way over the
-			// limit.
+		if httputil.NoNetwork(err) || httputil.CertExpiredOrNotValidYet(err) {
+			// If there is no network there is no need to
+			// count this as a tentatives attempt. If the
+			// cert is expired/not-valid yet that most
+			// likely means that the devices has no
+			// ntp-synced time yet.  No need to count that
+			// as a tentative attempt either.
```

Comment:
the issues of using the same retry for network and cert expired is that by definition with no network we are not bothering the service as we can't reach it but if for some reasons the cert expiry is not because of missed ntp sync or is non-recoverable we are now hammering the service. So I think we need a bit more careful here and decide on a slightly different strategy in the cert expired case. There is anyway some kind of pessimistic limit of how long ntp sync can really take, no? after the that anyway the backoff logic is no better or worse.

Reply:
Thanks, I looked at the systemd documentation and the max time between attempts to poll the NTP service is 2048s (34min8s). I changed the code to only allow 35 tentative attempts with ntp errors with a retry of 1min. This will ensure that the service is not hit too hard in case the cert is really expired. I did reuse `pre-poll-tentatives` though, I can change that and have it's own state var here but it seems unlikely that there is a mix or service related tentative errors and cert expired errors but I'm happy to change the code to keep the cert related attempts in their own var if you feel this is cleaner.